File: Shared/Blog.Core/Mapping/AutoMapperConfig.cs
// <copyright file="AutoMapperConfig.cs" company="Blog">
// Copyright (c) Blog. All rights reserved.
// </copyright>

namespace Blog.Core.Mapping
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Reflection;

    using AutoMapper;
    using Interfaces;

    /// <summary>
    /// AutoMapper Configuration.
    /// </summary>
    public static class AutoMapperConfig
    {
        /// <summary>
        /// Register mappings.
        /// </summary>
        /// <param name="assemblies">assemblies.</param>
        public static void RegisterMappings(params Assembly[] assemblies)
        {
            var types = assemblies.SelectMany(a => a.GetExportedTypes()).ToList();

            Mapper.Initialize(configuration =>
            {
                RegisterStandardFromMappings(configuration, types);

                RegisterStandardToMappings(configuration, types);

                RegisterCustomMaps(configuration, types);
            });
        }

        /// <summary>
        /// Register standard from mappings.
        /// </summary>
        /// <param name="configuration">configuration.</param>
        /// <param name="types">types.</param>
        private static void RegisterStandardFromMappings(IProfileExpression configuration, IEnumerable<Type> types)
        {
            var maps = GetFromMaps(types);

            CreateMappings(configuration, maps);
        }

        /// <summary>
        /// Register standard to mappings.
        /// </summary>
        /// <param name="configuration">configuration.</param>
        /// <param name="types">types.</param>
        private static void RegisterStandardToMappings(IProfileExpression configuration, IEnumerable<Type> types)
        {
            var maps = GetToMaps(types);

            CreateMappings(configuration, maps);
        }

        /// <summary>
        /// Register custom maps.
        /// </summary>
        /// <param name="configuration">configuration.</param>
        /// <param name="types">types.</param>
        private static void RegisterCustomMaps(IMapperConfigurationExpression configuration, IEnumerable<Type> types)
        {
            var maps = GetCustomMappings(types);

            CreateMappings(configuration, maps);
        }

        /// <summary>
        /// Get custom mappings.
        /// </summary>
        /// <param name="types">types.</param>
        /// <returns>IEnumerable.</returns>
        private static IEnumerable<ICustomMappings> GetCustomMappings(IEnumerable<Type> types)
        {
            var customMaps = from t in types
                             from i in t.GetTypeInfo().GetInterfaces()
                             where typeof(ICustomMappings).GetTypeInfo().IsAssignableFrom(t) &&
                                !t.GetTypeInfo().IsAbstract &&
                                !t.GetTypeInfo().IsInterface
                             select (ICustomMappings)Activator.CreateInstance(t);

            return customMaps;
        }

        /// <summary>
        /// Get from maps.
        /// </summary>
        /// <param name="types">types.</param>
        /// <returns>IEnumerable.</returns>
        private static IEnumerable<TypesMap> GetFromMaps(IEnumerable<Type> types)
        {
            var fromMaps = from t in types
                           from i in t.GetTypeInfo().GetInterfaces()
                           where i.GetTypeInfo().IsGenericType &&
                                 i.GetGenericTypeDefinition() == typeof(IMapFrom<>) &&
                                 !t.GetTypeInfo().IsAbstract &&
                                 !t.GetTypeInfo().IsInterface
                           select new TypesMap
                           {
                               Source = i.GetTypeInfo().GetGenericArguments()[0],
                               Destination = t,
                           };

            return fromMaps;
        }

        /// <summary>
        /// Get to maps.
        /// </summary>
        /// <param name="types">types.</param>
        /// <returns>IEnumerable.</returns>
        private static IEnumerable<TypesMap> GetToMaps(IEnumerable<Type> types)
        {
            var toMaps = from t in types
                         from i in t.GetTypeInfo().GetInterfaces()
                         where i.GetTypeInfo().IsGenericType &&
                               i.GetTypeInfo().GetGenericTypeDefinition() == typeof(IMapTo<>) &&
                               !t.GetTypeInfo().IsAbstract &&
                               !t.GetTypeInfo().IsInterface
                         select new TypesMap
                         {
                             Source = t,
                             Destination = i.GetTypeInfo().GetGenericArguments()[0],
                         };

            return toMaps;
        }

        /// <summary>
        /// Create mappings.
        /// </summary>
        /// <param name="configuration">configuration.</param>
        /// <param name="maps">maps.</param>
        private static void CreateMappings(IProfileExpression configuration, IEnumerable<TypesMap> maps)
        {
            foreach (var map in maps)
            {
                configuration.CreateMap(map.Source, map.Destination);
            }
        }

        /// <summary>
        /// Create mappings.
        /// </summary>
        /// <param name="configuration">configuration.</param>
        /// <param name="maps">maps.</param>
        private static void CreateMappings(IMapperConfigurationExpression configuration, IEnumerable<ICustomMappings> maps)
        {
            foreach (var map in maps)
            {
                map.CreateMappings(configuration);
            }
        }
    }
}

File: Shared/Blog.FSharp.Core/Mapping/AutoMapperConfig.fs
namespace Blog.FSharp.Core.Mapping

open System
open System.Reflection
open AutoMapper
open Blog.FSharp.Core.Mapping.Interfaces

/// AutoMapper Configuration
module AutoMapperConfig =

    /// Create standard mappings for TypesMap
    let private createMappings (configuration: IProfileExpression) (maps: TypesMap seq) =
        maps |> Seq.iter (fun map -> configuration.CreateMap(map.Source, map.Destination) |> ignore)

    /// Create custom mappings
    let private createCustomMappings (configuration: IMapperConfigurationExpression) (maps: ICustomMappings seq) =
        maps |> Seq.iter (fun map -> map.CreateMappings(configuration))

    /// Get "from" maps
    let private getFromMaps (types: Type list) : TypesMap seq =
        types
        |> Seq.filter (fun t -> not (t.GetTypeInfo().IsAbstract) && not (t.GetTypeInfo().IsInterface))
        |> Seq.collect (fun t ->
            t.GetTypeInfo().GetInterfaces()
            |> Seq.filter (fun i -> i.GetTypeInfo().IsGenericType && i.GetGenericTypeDefinition() = typedefof<IMapFrom<_>>)
            |> Seq.map (fun i -> 
                let map = TypesMap()
                map.Source <- i.GetGenericArguments().[0]
                map.Destination <- t
                map))

    /// Get "to" maps
    let private getToMaps (types: Type list) : TypesMap seq =
        types
        |> Seq.filter (fun t -> not (t.GetTypeInfo().IsAbstract) && not (t.GetTypeInfo().IsInterface))
        |> Seq.collect (fun t ->
            t.GetTypeInfo().GetInterfaces()
            |> Seq.filter (fun i -> i.GetTypeInfo().IsGenericType && i.GetGenericTypeDefinition() = typedefof<IMapTo<_>>)
            |> Seq.map (fun i ->
                let map = TypesMap()
                map.Source <- t
                map.Destination <- i.GetGenericArguments().[0]
                map))

    /// Get custom mappings
    let private getCustomMappings (types: Type list) : ICustomMappings seq =
        types
        |> Seq.filter (fun t -> not (t.GetTypeInfo().IsAbstract) && not (t.GetTypeInfo().IsInterface))
        |> Seq.filter (fun t -> typeof<ICustomMappings>.GetTypeInfo().IsAssignableFrom(t))
            |> Seq.map (fun t -> Activator.CreateInstance(t) :?> ICustomMappings)

    /// Register standard "from" mappings
    let private registerStandardFromMappings (configuration: IProfileExpression) (types: Type list) =
        let maps = getFromMaps types
        createMappings configuration maps

    /// Register standard "to" mappings
    let private registerStandardToMappings (configuration: IProfileExpression) (types: Type list) =
        let maps = getToMaps types
        createMappings configuration maps

    /// Register custom mappings
    let private registerCustomMaps (configuration: IMapperConfigurationExpression) (types: Type list) =
        let maps = getCustomMappings types
        createCustomMappings configuration maps

    /// Register mappings from assemblies
    let registerMappings ([<ParamArray>] assemblies: Assembly[]) =
        let types =
            assemblies
            |> Seq.collect (fun a -> a.GetExportedTypes())
            |> Seq.toList

        let config =
            MapperConfiguration(fun cfg ->
                registerStandardFromMappings cfg types
                registerStandardToMappings cfg types
                registerCustomMaps cfg types
            )

        config.CreateMapper() |> ignore
